packager/packager.h

```diff
@@ -0,0 +1,86 @@
+#ifndef SORBET_REWRITER_PACKAGE_H
+#define SORBET_REWRITER_PACKAGE_H
+#include "ast/ast.h"
+
+namespace sorbet {
+class WorkerPool;
+}
+
+namespace sorbet::packager {
+/**
+ * This pass transforms package files (`foo/__package.rb`) from
+ *
+ *   class Project::Foo < PackageSpec
+ *    import Project::Bar
+ *
+ *    export Baz
+ *    export_methods FooClassWithMethods
+ *   end
+ *
+ * given
+ *
+ *   class Project::Bar < PackageSpec
+ *     export SomeClassInBar
+ *     export Inner::SomeOtherClassInBar
+ *     export_methods SomeClassWithMethods, SomeOtherClassWithMethods
+ *   end
+ *
+ * to:
+ *
+ *   module <PackageRegistry>::Project_Foo_Package
+ *     module Project::Bar
+ *       # Import methods exported on Bar
+ *       extend <PackageRegistry>::Bar_Package::<PackageMethods>
+ *       # Import each class exported by Bar
+ *       SomeClassInBar = <PackageRegistry>::Bar_Package::SomeClassInBar
+ *       SomeOtherClassInBar = <PackageRegistry>::Bar_Package::SomeOtherClassInBar
+ *     end
+ *     module <PackageMethods>
+ *       include <PackageRegistry>::Project_Foo_Package::FooClassWithMethods
+ *     end
+ *   end
+ *
+ *   class Project::Foo < PackageSpec
+ *    import Project::Bar
+ *
+ *    export <PackageRegistry>::Project_Foo_Package::Baz
+ *    export_methods <PackageRegistry>::Project_Foo_Package::FooClassWithMethods
+ *   end
+ *
+ * It also rewrites files in the package, like `foo/baz.rb`, from:
+ *
+ * class SomeOtherClassInTheFile
+ *   ...
+ * end
+ *
+ * class Baz
+ *   ...
+ * end
+ *
+ * to:
+ *
+ * module <PackageRegistry>::Project_Foo_Package
```

Comment:
Is this mangling scheme sufficient to prevent naming conflicts? i.e., just replacing `::` with `_`? I'd imagine that it'd be pretty easy to have something like `class A::B::C` and `class A_B::C` collide with each other.

Reply:
We forbid `_` in package names.

Reply:
I can mention this in the comment as I imagine this will be a common concern.